### DISCUSSION TITLE
Parsing MyST in javascript by using mystjs as a module (not command)?

### DISCUSSION THREAD

**Discussion by lrq3000** [executablebooks/meta#1012] [Q&A] [answered]
Hello,

I would like to add MyST support in a multi-format editor (Markor on Android), and so I would need to be able to call mystjs as a simple javascript module to parse and render MyST syntax into HTML.

I am a quite experienced developer in Python and several other languages, except for javascript, so please excuse me if the following includes dumb questions 😅

There are two issues I am stumbling against to implement the above plan:
1. I am not sure how I can convert a npm package into an importable module directly in a javascript call. I found there are tools to do that, such as Browserify or webpack, is this correct?
2. Once it is bundled as an importable module, I am not sure how I can directly call mystjs various functions and submodules without using commandline commands, is there maybe some doc I missed? Is it possible at all?

Note that I tried to look into other attempts at making javascript editors for MyST, but there is only [one](https://github.com/antmicro/myst-editor) and it does not seem to use mystjs but just markdown-it.

Thank you very much in advance for your help!

**Comment by rowanc1** [comment, marked as answer]
Hi @lrq3000, thanks for your questions! There was some similar discussion here:
- https://github.com/orgs/executablebooks/discussions/981

And a video/getting-started that I made at that time here:
- https://github.com/rowanc1/myst-playground

Hopefully that starts to satisfy your use case. Happy to answer further questions as you dig in!

1. The libraries do generally need to be bundled to be useful in the browser or on the command line. In the future we will hopefully focus on out-of-the box libraries that can be used by adding a script tag to the page, but we haven't been focusing there quite yet.
2. There are a lot of different ways to use the various packages together in myst, these are various parsers, transformers and exporters. Hopefully the playground example gives you a hint of where to look. Other places to look are in this [demo](https://github.com/executablebooks/myst-theme/blob/main/packages/myst-demo/src/index.tsx) which powers the sandbox on myst-tools.org. That is based in react, but the parse command can show how some of the pieces and exporters (e.g. word, latex, etc) fit together.

There are some parts of the command line tools that don't currently transfer over to use directly in the browser quite yet (e.g. they have reliance on reading from disk to create a table of contents or read a file etc). We are looking to get that done as we are working towards the [jupyterlab-myst extension](https://github.com/executablebooks/jupyterlab-myst), which is also doing something similar to what you describe.

**Comment by lrq3000** [comment]
Thank you so much for all these pointers @rowanc1! I will try soon to make an implementation in Markor :-)

Also I wanted to ask you about the [mystjs](https://www.npmjs.com/package/mystjs) npm package, it seems to be an old release predating myst-cli, right?

**Reply by rowanc1** [reply]
The majority of that repository moved into `myst-parser` and `myst-to-html` see https://github.com/executablebooks/mystjs/issues/181 for a bit of an explanation of the rejig.

I think we may take over `mystjs` for a wrapper library in the future that allows a high-level interface similar to the CLI focused on parsing, that also works in the browser (the CLI doesn't as it is file-system focused in parts). We haven't got there yet though! Currently that library hasn't been released in a long time.
